src/typed-geometry/functions/objects/intersection.hh

```diff
@@ -2269,4 +2327,617 @@ template <class ScalarT>
+template <class ScalarT>
+[[nodiscard]] constexpr bool intersects(triangle<3, ScalarT> const& t, plane<3, ScalarT> const& plane)
+{
+    return intersects(plane, t);
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr optional<segment<3, ScalarT>> intersection(triangle<3, ScalarT> const& t1, triangle<3, ScalarT> const& t2)
+{
+    // early out: check with plane clamped by triangle t1
+    auto const plane_t1 = plane_of(t1);
+
+    if (!intersects(t2, plane_t1))
+        return {};
+
+    array<pos<3, ScalarT>, 2> insecs;
+    array<segment<3, ScalarT>, 3> segments_t1 = edges_of(t1);
+    array<segment<3, ScalarT>, 3> segments_t2 = edges_of(t2);
+    int insec_count = 0;
+
+    // check intersection of t1 segments with t2
+    for (auto const& s : segments_t1)
+    {
+        auto insec = intersection(s, t2);
+
+        if (insec.has_value())
+        {
+            insecs[insec_count++] = insec.value();
+        }
+
+        if (insec_count >= 2)
+            return segment<3, ScalarT>{insecs[0], insecs[1]};
+    }
+
+    // check intersection of t2 segments with t1
+    for (auto const& s : segments_t2)
+    {
+        auto insec = intersection(s, t1);
+        if (insec.has_value())
+        {
+            insecs[insec_count++] = insec.value();
+        }
+
+        if (insec_count >= 2)
+            return segment<3, ScalarT>{insecs[0], insecs[1]};
+    }
+
+    if (insec_count == 1)
+        return segment<3, ScalarT>{insecs[0], insecs[0]};
+
+    return {};
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr optional<pos<3, ScalarT>> intersection(segment<3, ScalarT> const& seg, triangle<3, ScalarT> const& t)
+{
+    dir<3, ScalarT> normal_t = normalize(cross((t.pos1 - t.pos0), (t.pos2 - t.pos0)));
+
+    plane<3, ScalarT> plane_t = plane<3, ScalarT>(normal_t, t.pos0);
+    // intersection point segment-plane
+    auto insec = intersection(seg, plane_t);
+    // early out
+    if (!insec.has_value())
+        return {};
+
+    // insec in triangle?
+    dir<3, ScalarT> a = normalize(cross(t.pos1 - t.pos0, normal_t));
+    dir<3, ScalarT> b = normalize(cross(t.pos2 - t.pos1, normal_t));
+    dir<3, ScalarT> c = normalize(cross(t.pos0 - t.pos2, normal_t));
+    bool b_a = signed_distance(insec.value(), plane<3, ScalarT>(a, t.pos1)) <= 0 ? false : true;
+    bool b_b = signed_distance(insec.value(), plane<3, ScalarT>(b, t.pos2)) <= 0 ? false : true;
+    bool b_c = signed_distance(insec.value(), plane<3, ScalarT>(c, t.pos0)) <= 0 ? false : true;
+
+    if (b_a == b_b && b_b == b_c)
+        return insec;
+
+    return {};
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr optional<pos<3, ScalarT>> intersection(triangle<3, ScalarT> const& t, segment<3, ScalarT> const& seg)
+{
+    return intersection(seg, t);
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr optional<segment<3, ScalarT>> intersection(segment<3, ScalarT> const& s, aabb<3, ScalarT> const& bb) // NOT CONFIRMED
+{
+    // TODO: There might be a more effective way!
```

Comment:
yeah there are a few good optimizations, please add a "TODO" comment above the function for that